machine-learning/03-Training and AutoML.ipynb

```diff
@@ -1,45 +1,561 @@
+        "\n",
+        "So what `Fit(x)` does is to lower the `loss`. A good model means a model with small `loss`, when you train a model, you want to decrease its loss so as to make the prediction of that model closer to actual value. That's what __Train__ actually mean."
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {},
+      "source": [
+        "## Trainers in ML.Net\n",
+        "There're plenty of trainers available in ML.Net, where you can find most of them under [StandardTrainersCatalog](https://docs.microsoft.com/en-us/dotnet/api/microsoft.ml.standardtrainerscatalog?view=ml-dotnet). Among those trainers, there're simple, linear ones like `SDCA`, `Lbfgs`, `LinearSvm`, or tree-based non-linear trainer, like `FastTree`, `RandomForest` and `LightGbm`. Generally, each trainer's capability is difference and non-linear models have a better training performance(lower loss) than linear ones, but it doesn't always mean they are always the better choice. It's an open question in picking trainers and usually requires many attempts to find out the best one.\n",
+        "\n",
+        "### Hyper-parameter optimization\n",
+        "Other than difference in trainers, different hyper-parameter in one trainer also have a huge impact over the final training performance, especially for tree-base trainers. This is because the capability of these trainers to fit a specific dataset is largly depends on their hyper parameters. For example, larger `numberOfLeaves` in `LightGbm` results to a larger model and usually enable it to fit on a more complex dataset, but it might have countereffect on small dataset and cause overfitting. On the contrary, if the dataset is complex but you set a small `numberOfLeaves`, it might impair `LightGbm`'s ability on fitting that dataset and cause underfit.\n",
+        "\n",
+        "In practice, it's usually tedious while necessary to try different set of hyper-parameters and find the best configuration for trainer, this process is called hyper-parameter optimization (HPO). Luckily, you can use the built-in `AutoML` to help you on hpo process.\n",
+        "\n",
+        "### OverFitting and UnderFitting\n",
+        "Overfitting and underfitting are the two most common problems we would see when training a model. UnderFitting means the selected trainer is not capable enough to fit training dataset and usually result in a high loss during training and low score/metric on test dataset. To resolve this we need either select a more powerful model, or do more feature engineering. And overfitting is just the opposite, which happens when model get overtrained and usually result in a decent low loss during training but low score on test dataset.\n",
```

Comment:
```suggestion
        "Overfitting and underfitting are the two most common problems you encounter when training a model. Underfitting means the selected trainer is not capable enough to fit training dataset and usually result in a high loss during training and low score/metric on test dataset. To resolve this you need to either select a more powerful model or perform more feature engineering. Overfitting is the opposite, which happens when model learns the training data too well. This usually results in low loss metric during training but high loss on test dataset.\n". 
        
        A good analogy for these concepts is studying for an exam. Let's say you knew the questions and answers ahead of time. After studying, you take the test and get a perfect score. Great news! However, when you're given the exam again with the questions rearranged and with slightly different wording you get a lower score. That suggests you memorized the answers and didn't actually learn the concepts you were being tested on. This is an example of overfitting. Underfitting is the opposite where the study materials you were given don't accurately represent what you're evaluated on for the exam. As a result, you resort to guessing the answers since you don't have enough knowledge to answer correctly.  
```